### ISSUE TITLE
RFC: Benchmarks between untraced and traced servers

### ISSUE THREAD

**Issue by bitwalker** [onetapbeyond/gen_metrics#1]
Hi! Great project!

I'm a bit concerned that the use of Erlangs tracing may be too significant a cost to pay for use with servers/stages which are heavily loaded, but I don't actually know for sure what the cost *is*. My suggestion I'd like you to consider is a quick benchmark comparing the speed at which a traced and untraced server, and a traced and untraced pipeline can perform the same work (can be something trivial like sending 1M messages containing a struct through them and see how long it takes to receive all of them or something).

I'd love to use this technique in my own application, but adding the tracing overhead is something I'm super hesitant to do. If I get a chance I'll try to report back with my own findings, but if you had a benchmark already set up in your project it'd be super easy to tweak it and test things based on my own workload.

Thanks again for sharing this project!

**Comment by gitpitch** [comment]
Cheers @bitwalker really appreciate this feedback and all your legitimate concerns so I plan to pull together some kind of (tweak-able) benchmarking tests to see what kinds of answers we can find to your questions. I'll update this thread once I have some data/code to share. 

Oops, posted as @gitpitch who also moonlights as @onetapbeyond

**Comment by bitwalker** [comment]
I've actually got a PR almost ready, I'll push it up tomorrow :)

**Comment by gitpitch** [comment]
Cool @bitwalker . I'll have something by then too so we can take a look tomorrow to see what makes most sense from our combined efforts.

**Comment by onetapbeyond** [comment]
Hi @bitwalker I see you have just submitted a PR which I will look tomorrow (it's v. late here now). In the meantime if you have a chance you might want to take a look at my own proposed solution to benchmarking GenMetrics, the new `gen_metrics_bench` package. 

The GitHub [repo is here](https://github.com/onetapbeyond/gen_metrics_bench) and the [HexDocs are here](https://hexdocs.pm/gen_metrics_bench). Most of the details are in the HexDocs and I have even included a few numbers which you may find interesting.

I haven't yet tested larger message size/types so as noted in the doc all data presented were captured running benchmarks using the default behaviours/configurations. As you will also see, you can easily plug in your own runtime-behaviours (message size, simulated task time) and see what affects those changes have using each of the available benchmark test harnesses for GenServer clusters and GenStage pipelines.

**Comment by bitwalker** [comment]
Hey @onetapbeyond, no worries! I took a look at your docs and the code, and here are my thoughts:

- There appears to be at least one significant flaw in your benchmark setup: once any of the `gen_metrics` monitors are activated, tracing is enabled *system wide* for all processes which use the same module/function/arg pattern, and since you are using the same module for each of the benchmarks, as soon as tracing is enabled once, it's enabled for all runs afterward (regardless of which benchmark is being run). I strongly suspect that's why your benchmark is showing no runtime impact, as I think you ran a trace at least once before running the "untraced" version. I'm *certain* that there is a performance overhead to using tracing, but that was never the problem, the question I'm most interested in is how severe is the overhead. For some people it will always be too much, but that's a given. Very low overhead would be great! However if the benchmark is showing no impact at all, we know for sure something is not right.
- It's difficult to understand __how__ the benchmark works internally. In the Elixir community we have a couple excellent benchmarking tools that are familiar enough to everyone that evaluating the quality of the benchmark is quite straightforward. I have to admit that I don't trust your benchmark implementation right now because the numbers just don't add up with what I'm seeing in mine, and there is very little to get wrong in mine. In addition, it does multiple runs, provides warm up time to make sure the code is hot before measuring, and then averages out the runs. Given how quick you turned this around, I'm assuming that you haven't had a chance to implement much of that yet.
- This is a pretty complex package for something which can be done in 100 lines of code with `benchee` or `benchfella`. I'm not saying it doesn't have potential to do more, or that it's a waste of time or something, just that as a consumer of libraries, I would much rather clone your `gen_metrics` repo, run the benchmarks with a few of my own custom tweaks (maybe even pull in some test data from my own application), and run it with `mix bench` than to try and figure out how to run `gen_metrics_bench` with my own custom modifications. This is in great part due to the fact that I'm already familiar with the use of those benchmarking tools, so they are quick and easy for me to jump in to.

To be clear, whichever choice you make is absolutely fine by me, I've had a chance to run my own benchmarks, and I need time to do more research which I don't have right now, so I'm going to table my idea for a bit, but I just wanted to follow up with you and give you my thoughts while it's still early enough to be useful to you. I hope it doesn't come across as exceedingly negative or offensive, as I think you've done some great work here :)

**Comment by onetapbeyond** [comment]
Hi @bitwalker I see nothing negative or offensive about any of this feedback, in fact it is exactly what I'm looking for. I readily acknowledge I have much to learn in the Elixir space so when you introduce me to popular go-to benchmarking tools or point out flaws in my own approach to benchmarking (for example, failing to disable tracing between runs) it's a great chance for me to learn and improve.

I'm just about to start a review of your actual PR so I'll update this thread later once I've got my head around the details and work out how best to proceed. Thanks again for all of your time and invaluable input so far.

**Comment by bitwalker** [comment]
No problem! If you have any questions, let me know and I'll make some time to answer :)

**Comment by onetapbeyond** [comment]
Hi @bitwalker so it has taken quite some time and it has been quite a journey for me  to be in a position now to be able to reconcile and explain how the benchmark numbers we both reported were so far apart.

### In brief:

1. The numbers I reported using my gen_metrics_bench library were in fact accurate.
2. The numbers you reported using your benchee benchmarks were also accurate :)
3. The numbers we reported were undoubtedly wildly different.
4. And that's ok, because I now know we were measuring wildly different things.

Rather than trying to summarize my findings I'd ask you to simply read them. I have created a detailed README for a new series of benchmarks (inspired by your own) that I have just pushed to the gen_metrics repo. You can find the benchmark [README here](https://github.com/onetapbeyond/gen_metrics/tree/master/bench).

I will not update the main README on the gen_metrics repo to link to this document until you have had a chance to review and respond to it's content.

### Some key context for you:

My own benchmarks differed from yours in two key, fundamental ways:

1. Your benchmarks used a synchronous (blocking) `GenStage.call/3` to push messages one-by-one into the pipeline
2. Your benchmarks also set `max_demand: 1` for your pipeline, essentially disabling batching of events when handling upstream demand.

My own benchmarks did not use synchronous calls at all. And my pipelines operated using the default `max_demand` configuration which dramatically changes the operating behaviour of the pipeline. Especially the `rate-of-callbacks-per-stage` in the pipeline that ultimately needs to be monitored for a given rate-of-throughput. This is something you will see discussed at length in the README I linked to.

### GenMetrics Library Update

Since this whole discussion began I have updated the GenMetrics library to disable the collection of metrics on synchronous calls by default. This feature can be re-enabled by using the `synchronous: true` option when you define your GenMetrics cluster or pipeline.

Basically, feeding millions of messages to a pipeline one message at a time via a synchronous `GenStage.call/3` call, introduces significant runtime load on GenMetrics with very little benefit if you are interested in monitoring the upstream performance of the actual stages within your pipeline. So rather than allowing this type of "drip-feeding" of a pipeline to define a benchmark result, I've now made it so it can be optionally included in the result timings. Again, this is something I cover in detail in the README.

Obviously there's lots here. As ever, I look forward to whatever feedback you can provide.

**Comment by bitwalker** [comment]
You're correct that my benchmark was setup not optimally, though that was by design - to truly test what the overhead is of the tracing, you need to set things up so that what you're measuring is exposed fully. By serializing calls (including in the pipeline), the time added by the tracing is much clearer. Of course in a real world situation you would architect things to use asynchrony when possible, but the cost calculated in my benchmark is still being paid, it's just hidden. Batching of course reduces the number of calls, making the impact of the tracing much less of an issue, but the point is to see the worst case, and to be able to build assumptions on top of that (i.e. if I'm 2x as slow in the worst case, then adjusting the batch size to be larger may make the difference moot).

I think what happened here is we were intentionally creating benchmarks for two different things - I was creating the benchmark for the worst case scenario so we could understand the semantics of running with tracing enabled, while you were creating the benchmark for a more typical scenario so that one could understand how performance was impacted in an optimally configured system. I missed the fact that you had set up casting vs calls and a different `max_demand`, so I assumed you were benchmarking the same thing I was - my bad :)

With the above in mind, I'm not sure that disabling tracing of synchronous callbacks actually is a good idea. You still pay the cost on async message handling, it's just not visible to the consumer (although that's not true either if the consumer `cast` a message, but then enters a receive block, expecting an asynchronous reply from the server, which is a common pattern). Basically, I think that if you are going to turn tracing on, it should be fully on. I doubt many people are going to want to skip metrics on calls. That's just my opinion though.

**Comment by onetapbeyond** [comment]
Hi @bitwalker thanks for making time to look this over and again providing very constructive feedback.

I think the fact that we both took different approaches to the design of our benchmarks is highly fortuitous because it has forced me to go back and answer important questions about worst case v typical case scenarios. And I hope you agree that the [README](https://github.com/onetapbeyond/gen_metrics/tree/master/bench) presents this data, both good news and bad news, without bias.

Your original question that launched this thread was as follows:

> I'm a bit concerned that the use of Erlangs tracing may be too significant a cost to pay for use with servers/stages which are heavily loaded, but I don't actually know for sure what the cost is...I'd love to use this technique in my own application, but adding the tracing overhead is something I'm super hesitant to do.

Ignoring the [TCO issue](https://github.com/onetapbeyond/gen_metrics/issues/3) for now (don't worry I'm not, it's next on my list), given what you know about the design of your own application and what we now know about GenMetrics runtime costs under different conditions, are you any less hesitant to activate GenMetrics for your application?

Finally, as to whether disabling of tracing to synchronous calls is a good idea. Having rethought this I must agree with your take on this: if you are going to turn tracing on, it should be fully on. This should indeed be the default behaviour. I'm going to make that change, while preserving an option so people can disable synchronous callbacks if needed.

**Comment by bitwalker** [comment]
I would agree that your README is unbiased and well presented :)

Regarding the use of GenMetrics in my own application - and TCO issue aside - I'm still hesitant, but that's only because the performance characteristics of my application leave me with tight margins before overhead impacts throughput significantly. The application has multiple complex GenStage pipelines, as well as lots of standalone service-like processes, often pools of them, where events come in to the system at rates of up to 1k/second, but trigger workflows which may produce 10s of additional events for each input event. This means that pretty much every server-like process in the system is dealing with an almost non-stop stream of messages, the GenStage pipelines are configured to maximize throughput but balance that with restricting the degree of parallelism in the system, so the batches are smaller, many times in the range of 20-50 events per batch, so the overhead would be more noticeable with my GenStage pipelines than they would be in some other cases.

That said, my application is a complex beast - instrumentation without inducing a lot of overhead is already a tricky problem. For example, I've had to implement the sampling approach for some particularly hot areas of code, where metrics are critical to anticipating problems, but we can't afford to generate datapoints every time the code is hit, so instead we only generate the data maybe 5% of the time or 10% of the time. This is why I recommended this approach to you, because it is suitable for environments where metrics are important but so is performance - sampling rather than always collecting will mean you can afford to go deeper with what you collect without impacting the system too negatively.

**Comment by onetapbeyond** [comment]
Really appreciate all of the background info, it provides great context as I think about real world usage of GenMetrics. That really is some application you are building, a complex beast indeed.

Sampling is a great suggestion and I can see how it can solve a number of problems and broaden the usefulness of any metrics gathering library, including GenMetrics. Work on this has begun. As ever, I'll update this thread once I've something to show that may be of interest.

**Comment by onetapbeyond** [comment]
GenMetrics now supports metrics sampling. The project [README](https://github.com/onetapbeyond/gen_metrics), [HexDocs](https://hexdocs.pm/gen_metrics/GenMetrics.html), and [GitPitch](https://gitpitch.com/onetapbeyond/gen_metrics) have been updated to reflect these changes. I have also completely rewritten the [benchmarks performance guide](https://github.com/onetapbeyond/gen_metrics/tree/master/bench) as it now highlights the runtime benefits of using sampling when monitoring servers and pipelines.

@bitwalker you might be very interested in the overall impact sampling has on the benchmark results. That said, the README, the HexDocs and the benchmarks guide all still include a warning about using GenMetrics in long-running production environments. These recent changes can significantly delay resource exhaustion, but with enough time the inevitable is, well, inevitable.

If you have a chance to look this over I'm still wide open to any feedback or suggestions. Thanks.

**Comment by bitwalker** [comment]
Thanks again for your patience :)
